File: backend/ml-service/glassdoor_service.py
"""
Glassdoor service for fetching company reviews and ratings
Note: This is a placeholder implementation. In production, consider using:
1. Official Glassdoor API (if available and approved)
2. Licensed third-party data providers
3. More sophisticated scraping with proper rate limiting and error handling
"""
import logging
import os
import requests
from typing import Dict, Any
import json
import time

logger = logging.getLogger(__name__)

class GlassdoorService:
    def __init__(self):
        """Initialize Glassdoor service"""
        self.api_key = os.environ.get('GLASSDOOR_API_KEY')
        self.partner_id = os.environ.get('GLASSDOOR_PARTNER_ID')
        self.enabled = bool(self.api_key and self.partner_id)

        if self.enabled:
            logger.info("Glassdoor service initialized with API credentials")
        else:
            logger.info("Glassdoor service running in limited mode (no API credentials)")

    def is_available(self) -> bool:
        """Check if Glassdoor service is configured (always returns True for basic functionality)"""
        return True  # Always available, but may have limited functionality without API key

    def get_company_rating(self, company_name: str) -> Dict[str, Any]:
        """
        Get company rating and sentiment from Glassdoor

        Args:
            company_name: Name of the company to look up

        Returns:
            Dictionary with rating (0-5 scale), sentiment (-1 to 1), confidence, and review count
        """
        try:
            # If we have API credentials, try to use the official API
            if self.enabled:
                return self._get_rating_via_api(company_name)
            else:
                # Fallback: return neutral response
                # In a production system, you might implement web scraping here
                # with proper rate limiting, user-agent rotation, and respect for robots.txt
                return self._get_default_rating()

        except Exception as e:
            logger.error(f"Error getting Glassdoor rating for {company_name}: {e}")
            return self._get_default_rating()

    def _get_rating_via_api(self, company_name: str) -> Dict[str, Any]:
        """
        Get rating using Glassdoor API (placeholder implementation)
        Note: Actual Glassdoor API implementation would require partnership approval
        """
        try:
            # This is a placeholder for the actual Glassdoor API call
            # The real Glassdoor API endpoint would look something like:
            # http://api.glassdoor.com/api/api.htm
            #
            # Parameters would include:
            # - v: '1'
            # - format: 'json'
            # - t.p: partner_id
            # - t.k: api_key
            # - userip: '0.0.0.0'
            # - useragent: 'YourApp/1.0'
            # - action: 'employers'
            # - q: company_name

            # For now, we'll simulate a delay and return structured data
            # In a real implementation, you would make the actual HTTP request here

            # Simulate API delay
            time.sleep(0.1)

            # Placeholder response - in reality, this would come from the API
            return {
                'rating': 3.5,  # 0-5 scale
                'sentiment': 0.2,  # -1 to 1 scale (derived from rating)
                'confidence': 0.7,  # Based on having API access
                'review_count': 42  # Placeholder
            }

        except Exception as e:
            logger.warning(f"Glassdoor API request failed: {e}")
            return self._get_default_rating()

    def _get_default_rating(self) -> Dict[str, Any]:
        """
        Return a default/rating when Glassdoor data is not available
        This ensures the system continues to function even without Glassdoor access
        """
        return {
            'rating': 0.0,      # Indicates no data available
            'sentiment': 0.0,   # Neutral sentiment
            'confidence': 0.0,  # No confidence without data
            'review_count': 0   # No reviews found
        }

    def extract_review_sentiment(self, review_text: str) -> float:
        """
        Extract sentiment from review text
        Returns score between -1 (very negative) and 1 (very positive)
        """
        if not review_text or not isinstance(review_text, str):
            return 0.0

        # Similar sentiment analysis as Reddit service
        text_lower = review_text.lower()

        positive_indicators = [
            'recommend', 'recommended', 'great', 'excellent', 'good', 'best',
            'positive', 'happy', 'satisfied', 'worth it', 'valuable', 'learning',
            'growth', 'opportunity', 'flexible', 'benefits', 'culture', 'team',
            'management', 'leadership', 'work-life balance', 'compensation',
            'proud', 'proud to work', 'proud to be', 'great place', 'great company',
            'amazing', 'fantastic', 'wonderful', 'terrific', 'outstanding',
            'superb', 'brilliant', 'exceptional', 'outstanding', 'first-rate',
            'top-notch', 'world-class', 'leading', 'innovative', 'dynamic'
        ]

        negative_indicators = [
            'not recommend', 'avoid', 'terrible', 'awful', 'bad', 'worst',
            'disappointed', 'unhappy', 'miserable', 'toxic', 'stressful',
            'overworked', 'underpaid', 'poor management', 'no growth',
            'layoffs', 'fired', 'quit', 'resigned', 'stress', 'burnout',
            'micromanaged', 'favoritism', 'discrimination', 'harassment',
            'hostile', 'hostile work environment', 'unsafe', 'unprofessional',
            'disorganized', 'chaotic', 'messy', 'dirty', 'broken', 'broken system',
            'broken promises', 'broken culture', 'broken leadership', 'broken management'
        ]

        positive_count = sum(1 for word in positive_indicators if word in text_lower)
        negative_count = sum(1 for word in negative_indicators if word in text_lower)

        total = positive_count + negative_count
        if total == 0:
            return 0.0

        return (positive_count - negative_count) / max(total, 1)

# Singleton instance
glassdoor_service = GlassdoorService()

def get_company_glassdoor_rating(company_name: str) -> dict:
    """
    Convenience function to get Glassdoor rating for a company

    Args:
        company_name: Name of the company

    Returns:
        Dictionary with rating analysis results
    """
    return glassdoor_service.get_company_rating(company_name)

def is_glassdoor_available() -> bool:
    """
    Check if Glassdoor service is available

    Returns:
        True (always available, though may have limited functionality without API key)
    """
    return glassdoor_service.is_available()

File: backend/src/services/glassdoorService.js
import axios from "axios";
import { env } from "../config/env.js";

/**
 * Glassdoor service for fetching company reviews and ratings
 * Note: Glassdoor has strong anti-bot measures, so this is a simplified implementation
 * In production, you might want to use official APIs or more sophisticated scraping techniques
 */
export async function getCompanyReviews(companyName, limit = 10) {
  try {
    // Note: Direct scraping of Glassdoor is challenging due to anti-bot measures
    // This is a placeholder implementation that returns structured data
    // For a real implementation, consider:
    // 1. Using Glassdoor's official API if available
    // 2. Using third-party services that provide Glassdoor data
    // 3. Implementing more sophisticated scraping with rate limiting and proxies

    // For now, we'll return empty results and rely on other data sources
    // This maintains compatibility while allowing for future enhancement

    // Try to get data from alternative sources if available
    try {
      // Attempt to use a general search API or service if configured
      if (env.GLASSDOOR_API_KEY) {
        const response = await axios.get('https://api.glassdoor.com/api/api.htm', {
          params: {
            v: '1',
            format: 'json',
            t.p: 'YOUR_PARTNER_ID',
            t.k: 'YOUR_API_KEY',
            userip: '0.0.0.0',
            useragent: '"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"',
            action: 'employers',
            q: companyName
          },
          timeout: 5000
        });

        if (response.data && response.data.response &&
            response.data.response.employers &&
            response.data.response.employers.length > 0) {

          const employer = response.data.response.employers[0];
          return {
            rating: employer.overallRating || 0,
            reviewCount: employer.numberOfRatings || 0,
            reviews: [] // Would contain actual review texts in a full implementation
          };
        }
      }
    } catch (apiError) {
      // Fall through to return empty result
      console.warn('[GlassdoorService] API attempt failed:', apiError.message);
    }

    // Return empty structure - no Glassdoor data available
    return {
      rating: 0,
      reviewCount: 0,
      reviews: []
    };
  } catch (error) {
    console.error('[GlassdoorService] Error fetching reviews:', error.message);
    return {
      rating: 0,
      reviewCount: 0,
      reviews: []
    };
  }
}

/**
 * Extract sentiment from review text
 * Returns score between -1 (very negative) and 1 (very positive)
 */
export function extractReviewSentiment(reviewText) {
  if (!reviewText) return 0;

  // Similar sentiment analysis as Reddit service but tuned for reviews
  const positiveIndicators = [
    'recommend', 'recommended', 'great', 'excellent', 'good', 'best',
    'positive', 'happy', 'satisfied', 'worth it', 'valuable', 'learning',
    'growth', 'opportunity', 'flexible', 'benefits', 'culture', 'team',
    'management', 'leadership', 'work-life balance', 'compensation'
  ];

  const negativeIndicators = [
    'not recommend', 'avoid', 'terrible', 'awful', 'bad', 'worst',
    'disappointed', 'unhappy', 'miserable', 'toxic', 'stressful',
    'overworked', 'underpaid', 'poor management', 'no growth',
    'layoffs', 'fired', 'quit', 'resigned', 'stress', 'burnout',
    'micromanaged', 'favoritism', 'discrimination', 'harassment'
  ];

  const textLower = reviewText.toLowerCase();
  let positiveScore = 0;
  let negativeScore = 0;

  positiveIndicators.forEach(word => {
    if (textLower.includes(word)) {
      positiveScore += 1;
    }
  });

  negativeIndicators.forEach(word => {
    if (textLower.includes(word)) {
      negativeScore += 1;
    }
  });

  const total = positiveScore + negativeScore;
  if (total === 0) return 0;

  return (positiveScore - negativeScore) / Math.max(total, 1);
}

/**
 * Get aggregated sentiment and rating for a company from Glassdoor
 */
export async function getCompanyRating(companyName) {
  try {
    const result = await getCompanyReviews(companyName);

    // Convert rating (typically 0-5 scale) to -1 to 1 scale for sentiment
    // Assume 3.0 is neutral (0 sentiment), 5.0 is very positive (1.0), 1.0 is very negative (-1.0)
    const normalizedRating = (result.rating - 3) / 2; // Maps 1-5 to -1 to 1

    // Analyze sentiment from review texts if available
    let reviewSentiment = 0;
    if (result.reviews && result.reviews.length > 0) {
      const sentiments = result.reviews.map(review =>
        extractReviewSentiment(review.text || review.summary || '')
      );
      const validSentiments = sentiments.filter(s => !isNaN(s));
      if (validSentiments.length > 0) {
        reviewSentiment = validSentiments.reduce((sum, val) => sum + val, 0) / validSentiments.length;
      }
    }

    // Combine rating sentiment and review sentiment (weighted average)
    // Give more weight to the numerical rating as it's more reliable
    const combinedSentiment = (normalizedRating * 0.7) + (reviewSentiment * 0.3);

    // Confidence based on number of reviews
    const confidence = Math.min(1.0, result.reviewCount / 50); // Max confidence at 50+ reviews

    return {
      rating: result.rating, // Original 0-5 scale
      sentiment: Math.max(-1, Math.min(1, combinedSentiment)), // Ensure -1 to 1 range
      confidence: confidence,
      reviewCount: result.reviewCount
    };
  } catch (error) {
    console.error('[GlassdoorService] Error getting company rating:', error.message);
    return {
      rating: 0,
      sentiment: 0,
      confidence: 0,
      reviewCount: 0
    };
  }
}